### ISSUE TITLE
Isodata

### ISSUE THREAD

**Issue by vavrecan** [preactjs/preact-iso#38]
Hello, why is isodata created?
https://github.com/preactjs/wmr/issues/592
https://github.com/preactjs/preact-iso/blob/main/src/prerender.js#L37
Thank you

**Comment by rschristian** [comment]
Right now, it's primarily for `hydrate` to determine whether to hydrate or render.

As mentioned, we had some discussions internally a few years back, I'm not sure if we ever fully decided on what we wanted to do.

**Comment by vavrecan** [comment]
It should be defined as `text/isodata` by default, with an option to make
it configurable for greater flexibility.

**Comment by rschristian** [comment]
May I ask why? And what should be configurable?

**Comment by rschristian** [comment]
> The "subtype" must always be included in the "type," just as "mimetype" is required

That's more of a suggestion than a legitimate requirement. This is pretty standard, for good or bad.

**Comment by vavrecan** [comment]
type without subtype/ prefix will fail w3 validator with error not warning

**Comment by vavrecan** [comment]
Thank you for the clarification! I ended up writing my own prerender and hydrate methods. This allowed me to store the props data directly in <script> tags, so now all props are available in the HTML. This way, the data props don't have to be loaded from external js, and I can modify them server-side if needed

 <div id="app"><div>Hello world</div><script type="application/json" id="isodata" data-route="/product">{"title":"hello world"}</script></div>

**Comment by rschristian** [comment]
Glad to hear it!

> It would be great if the library had an option to save props data into isodata script tags, so this could be done more easily.

We'll probably support that via `@preact/preset-vite` instead, in the same API we used for WMR. I don't think that's something this, as a thin wrapper around RTS, necessarily needs to be aware of.

**Comment by vavrecan** [comment]
That's perfect! I'm also using `preact/preset-vite` atm

**Comment by rschristian** [comment]
Great!

I didn't originally copy it over as I wasn't quite sure where we wanted it (especially with `preact-iso` being separated out, as it was more tightly-bound to WMR), but never circled back around to that. Will try to get it added soon though. It'll be a `data` key on the return object, i.e.,

```js
export async function prerender(data) {
    ...
    return { html, data: { title: 'Hello World!' } };
}
```

**Comment by rschristian** [comment]
The invalid script type is [fixed for v3](https://github.com/preactjs/preact-iso/pull/17/commits/f2b76b5668a840be08e91a0b5c22ecc90661b2cb), switched it to `<script id="isodata">`.

Prerender data provider for `@preact/preset-vite` landed [here](https://github.com/preactjs/preset-vite/pull/141) and is available in v2.9.1

Thanks for opening this!
